dali/pipeline/operator/builtin/external_source.h

```diff
@@ -215,25 +167,36 @@ class ExternalSource : public Operator<Backend> {
 
  protected:
   bool SetupImpl(std::vector<OutputDesc> &output_desc, const workspace_t<Backend> &ws) override {
+    if (no_copy_) {
+      return false;
+    }
     TensorListShape<> shape;
     output_desc.resize(1);
     {
       std::unique_lock<std::mutex> busy_lock(busy_m_);
       if (blocking_) {
-        cv_.wait(busy_lock, [&data = tl_data_] {return !data.IsEmpty(); });
+        cv_.wait(busy_lock, [&data = state_] {return !data.empty(); });
       } else {
-        if (tl_data_.IsEmpty()) {
+        if (state_.empty()) {
           DALI_FAIL("No data was provided to the ExternalSource. Make sure to feed it properly.");
         }
       }
-      output_desc[0].shape = tl_data_.PeekFront()->shape();
-      output_desc[0].type = tl_data_.PeekFront()->type();
+      state_.pop_front();
+      // for zero copy we don't want any shape inference to avoid any data allocation ahead
```

Comment:
I guess this line was meant to go to line 170 (?)

Reply:
Pre-rework leftover.

Reply:
Done